### PR TITLE
Nested groups support

### PR DESCRIPTION
**Pull request by marcocaberletti**
This PR privides a first implementation of nested groups, as described in #81 .
Main changes:
- new develop version;
- ScimGroup extension that introduces parent group;
- new database migration to add parent group foreign key;
- unit tests for IamGroup JPA repository;
- fixes to ScimGroupProvisioning service to manage parent group in create/delete.

### CODE REVIEW

File: iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    ScimGroup animals = createGroup("animals");
+    ScimGroup mammals = createGroup("mammals", animals);
+
+    // @formatter:off
+    mvc.perform(get("/scim/Groups/{id}", mammals.getId()))
+      .andDo(print())
+      .andExpect(status().isOk())
+      .andExpect(jsonPath("$.id", equalTo(mammals.getId())))
+      .andExpect(jsonPath("$.displayName", equalTo(mammals.getDisplayName())))
+      .andExpect(jsonPath("$."+ScimConstants.INDIGO_GROUP_SCHEMA+".parentGroup.display", equalTo(animals.getDisplayName())))
```
Comment (andreaceccanti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java]: Stuff from ScimConstants can be statically imported to save some space

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java [hunk 2]
```diff
@@ -105,9 +126,18 @@ public void delete(String id) {
...
+    if (!(group.getAccounts().isEmpty() && group.getChildrenGroups().isEmpty())) {
+
+      throw new ScimException("Group is not empty");
+    }
+
+    IamGroup parent = group.getParentGroup();
+    if (parent != null) {
+      Set<IamGroup> children = parent.getChildrenGroups();
+      children.remove(group);
+      parent.setChildrenGroups(children);
```
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  @Test
+  public void createParentGroup() {
+
+    parent = createGroup(null);
+
+    IamGroup group = groupRepository.findByUuid(parent.getUuid()).get();
+    assertNotNull(group);
+    assertNull(group.getParentGroup());
+    assertThat(group.getChildrenGroups(), Matchers.empty());
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    parent = createGroup(null);
+    child = createGroup(parent);
+
+    IamGroup group = groupRepository.findByUuid(child.getUuid()).get();
+    assertNotNull(group.getParentGroup());
+    assertEquals(parent.getUuid(), group.getParentGroup().getUuid());
+
+    group = groupRepository.findByUuid(parent.getUuid()).get();
+    assertThat(group.getChildrenGroups(), Matchers.not(Matchers.empty()));
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    IamGroup group = groupRepository.findByUuid(child.getUuid()).get();
+    groupRepository.delete(group);
+    Set<IamGroup> children = parent.getChildrenGroups();
+    children.remove(child);
+    parent.setChildrenGroups(children);
+
+    groupRepository.save(parent);
+
+    group = groupRepository.findByUuid(parent.getUuid()).get();
+    assertThat(group.getChildrenGroups(), Matchers.empty());
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    parent = createGroup(null);
+    assertEquals(count + 1, groupRepository.findRootGroups().size());
+  }
+
+  @Test
+  public void listSubgroups() {
+    parent = createGroup(null);
+    List<IamGroup> subgroups = groupRepository.findSubgroups(parent);
+    assertThat(subgroups, Matchers.empty());
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/repository/IamGroupRepositoryTests.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+  @Autowired
+  private ScimResourceLocationProvider scimResourceLocationProvider;
+
+  private MockMvc mvc;
+  private ObjectMapper objectMapper;
+
+  @Before
+  public void setup() {
+    mvc = MockMvcBuilders.webAppContextSetup(context).apply(springSecurity()).build();
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java]: Done.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    objectMapper = JacksonUtils.createJacksonObjectMapper();
+  }
+
+  @Test
+  @WithMockOAuthUser(clientId = "scim-client-rw", scopes = {"scim:read", "scim:write"})
+  public void testCreateNewChildGroup() throws Exception {
+
+    ScimGroup animals = createGroup("animals");
+    createGroup("mammals", animals);
```
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java]: Added some assertions.

File: iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java (hunk 1 above)
Comment (marcocaberletti) [inline iam-login-service/src/test/java/it/infn/mw/iam/test/scim/group/ScimGroupTests.java]: Done.

**Pull request merged**